fix(Select): mutual exclusive props

nubsthead opened the pull request:
Select component can distinguish all the use cases and required props:
- single selection controlled mode
- single selection uncontrolled mode
- multiple selection controlled mode
- multiple selection uncontrolled mode

nubsthead committed:
fix(Select): mutual exclusive props

Select component can distinguish all the use cases and required props:
- single selection controlled mode
- single selection uncontrolled mode
- multiple selection controlled mode
- multiple selection uncontrolled mode

nubsthead commented:
With this implementation all the required props will depend on the type of usage of the select. 
Examples:
- You are not allowed to pass both selection and defaultSelection, only one can be provided
- You are not allowed to use onChange(arg: string | null) => void if multiple is set to true; onChange(arg: SelectItem[]) => void must be provided
- You are not allowed to provide (selection or defaultSelection): SelectItem if multiple is set to true; SelectItem[] is allowed

beawar commented on src/components/inputs/Select.tsx:
```diff
@@ -254,22 +273,28 @@ const Select = React.forwardRef<HTMLDivElement, SelectProps>(function SelectFn(
+			(onChange as SingleSelectionOnChange)?.(selected[0]?.value ?? null);
+		}
 		// eslint-disable-next-line react-hooks/exhaustive-deps
 	}, [selected, multiple]);
```
I think we should remove this eslint comment and add onChange as dependency of the hook, or otherwise we could incur into unwanted behaviours. What do you think?

nubsthead replied:
I agree with and I would like to remove it as well but at the current development stage of the Select component we can't add the onChange to the dependencies as it will create a memory leak inside the apps.
If you want I can add a comment over it to keep track of this issue

beawar replied:
You are right. A comment sounds good, so that we can easily remember why we didn't remove it and also that it needs a deeper study/refactor.

nubsthead replied:
Done

nubsthead committed:
chore(Select): comment on useEffect without exhaustive deps

beawar reviewed (dismissed):
LGTM

nubsthead commented:
There is actually another noticeable change on this PR.
The onChange is now dynamically typed depending on Select usage mode (multiple selection, single selection).
Some types around the app could now be considered wrong: (e.g. below)

`onChange: <NonNullable<SelectProps['onChange']>>`

This will be considered error because `SelectProps['onChange']` will return type of `MultipleSelectionOnChange | SingleSelectionOnChange`.
The onChange must be either one of the type `MultipleSelectionOnChange` or `SingleSelectionOnChange` depending from the usage (multiple or not) and not both.

beawar commented:
> There is actually another noticeable change on this PR. The onChange is now dynamically typed depending on Select usage mode (multiple selection, single selection). Some types around the app could now be considered wrong: (e.g. below)
> 
> `onChange: <NonNullable<SelectProps['onChange']>>`
> 
> This will be considered error because `SelectProps['onChange']` will return type of `MultipleSelectionOnChange | SingleSelectionOnChange`. The onChange must be either one of the type `MultipleSelectionOnChange` or `SingleSelectionOnChange` depending from the usage (multiple or not) and not both.

Then the two types should be exported to be easily usable from the outside. Add them to the export on the end of the file to include them in the types bundled in the final package. I think a couple of test are needed to check these situations immediately from within the project itself.

nubsthead commented:
> > There is actually another noticeable change on this PR. The onChange is now dynamically typed depending on Select usage mode (multiple selection, single selection). Some types around the app could now be considered wrong: (e.g. below)
> > `onChange: <NonNullable<SelectProps['onChange']>>`
> > This will be considered error because `SelectProps['onChange']` will return type of `MultipleSelectionOnChange | SingleSelectionOnChange`. The onChange must be either one of the type `MultipleSelectionOnChange` or `SingleSelectionOnChange` depending from the usage (multiple or not) and not both.
> 
> Then the two types should be exported to be easily usable from the outside. Add them to the export on the end of the file to include them in the types bundled in the final package. I think a couple of test are needed to check these situations immediately from within the project itself.

I completely agree and I was already evaluating a similar solution. As this task will take some time to be considered completed I will convert this PR to draft and implement a proper solution later in a different sprint.

nubsthead committed:
refactor(Select): refactor onChange behaviour

Select onChange is now called only when slected value(s) differs from previous value.
It also fix controlled mode, now its value during controlled mode can only change from outside.

nubsthead committed:
chore: code cleanup

nubsthead committed:
fix: controlled mode - triggering internal state changes only when selection changes

nubsthead committed:
test(Select): select usage test implementation

nubsthead committed:
chore: removing logs from code

beawar commented on src/components/inputs/Select.test.tsx:
```diff
@@ -0,0 +1,315 @@
+			expect(onChange).not.toHaveBeenCalled();
+
+			// clicking on the Select renders the dropdown items
+			await userEvent.click(screen.getByText(label));
```
fyi: the await is not necessary since we are still using jest 26 (but you can leave it there, I want to migrate to jest 28/29 soon)

nubsthead replied:
You're right, I didn't notice the dependency version

beawar commented on src/components/inputs/Select.md:
```diff
@@ -45,23 +47,27 @@ const items = [
 ];
-const [selected, setSelected] = useState(4);
+const [selected, setSelected] = useState();
+const onChange = (ev) => {
```
newValue instead of ev? ev recalls to the event, but this is not the case

beawar commented on src/components/inputs/Select.md:
```diff
@@ -45,23 +47,27 @@ const items = [
+		onChange={onChange}
 	/>
-	<Text>Currently selected: {items[selected - 1].label}</Text>
+    <Text>Currently selected: {selected ? items[selected - 1].label : 'no selection'}</Text>
```
Selected is a string. There is a type coercion here that is dangerous imo (there was also before in other examples, but it might be right moment to fix it since we are refactoring)

beawar reviewed (commented):
I'm a little confused about the test cases. I mean, we could improve the description of the test cases and then split the tests to check only the rendering and then only the behaviour. What is confusing right now is that for the same format of the description we have different formats of the test. For example in _controlled mode - no default selection_ the test check also the click behaviour, while this is not happening for the _controlled mode - default selection_ test. I think these could be instead grouped inside a _description('Controlled mode')_ and then split into
 * test('if there is a default selection, both the label and the selection are visible. On change is not called')
 * test('if there is no default selection, only the label is visible')
 * test('if there is a default selection, onChange is called if the user clicks on an item with a different value')
 * test('if there is a default selection, onChange is not called if the user clicks on the item with the same value')
 * test('if there is no default selection, onChange is called if the user clicks on the item')
 * test('click on an item in controlled mode does not automatically update the value of the select')
 * ... and so on

See how many single cases are emerging for this simple combination of props.

Also, I think we should not use auxiliary component if not strictly necessary. We can instead use the rerender function returned by the render function, which allows us to rerender the tested component with different props.
For example, for the controlled component we could implement the following test
```tsx
test('If the value change, the new value is shown as the one selected', () => {
  const label = 'Select an item';
  const onChange = jest.fn();
  const { rerender } = render(
    <Select items={items} label={label} onChange={onChange} selection={items[0]} />
  );
  expect(screen.getByText(label)).toBeVisible();
  expect(screen.getByText(items[0].label)).toBeVisible();
  rerender(<Select items={items} label={label} onChange={onChange} selection={items[1]} />);
  expect(screen.queryByText(items[0].label)).not.toBeInTheDocument();
  expect(screen.getByText(items[1].label)).toBeVisible();
});
```

nubsthead committed:
fix(test): reformatting tests, add new test cases, fixes on select

beawar commented on src/components/display/Dropdown.tsx:
```diff
@@ -140,6 +140,7 @@ function PopperListItem({
 	return (
 		<ContainerEl
 			ref={itemRef}
+			data-testid={label}
```
this is not necessary. If you need a data-testid, you can specify it in the item object

beawar commented on src/components/inputs/Select.test.tsx:
```diff
@@ -0,0 +1,328 @@
+			await userEvent.click(screen.getByTestId(extendedItems[5].label));
+
+			expect(onChange).not.toBeCalled();
+			expect(screen.getByTestId(extendedItems[0].label)).toBeVisible();
```
Why not getByText instead of getByTestId? If the item should be visible, the text label should be visible too.
In this case, correct me if I'm wrong, the problem is there are two strings for the same label visible. What do you want to check with this expect?

nubsthead replied:
Yes you are correct. There are two strings with the same text, one is the label of the select and one is the label of the dropdown option. I wanted to make sure to avoid conflicts between those two adding a test id to the dropdown options but I am open to cleaner alternatives.

In this case I want the label of the select to be the same as the label of the dropdown option

beawar replied:
So, in order to follow what the user would see, we can check that there is only 1 label for the disabled option and 2 (and 2 only) labels visible for the selected option. Transformed in code:
```ts
expect(screen.getByText(extendedItems[5].label)).toBeVisible();
const selectedItemLabels = screen.getAllByText(extendedItems[0].label);
expect(selectedItemLabels).toHaveLength(2);
expect(selectedItemLabels[0]).toBeVisible();
expect(selectedItemLabels[1]).toBeVisible();
```

beawar commented on src/components/inputs/Select.test.tsx:
```diff
@@ -0,0 +1,328 @@
+
+describe('Select', () => {
+	const label = 'Select an item';
+	const onChange = jest.fn();
```
I think it's better if this mock is initialized inside every single test, to avoid possible conflicts

nubsthead replied:
Yes, that is true, I will move them all inside every single test

nubsthead committed:
fix: using correct selectors without adding a new testId

nubsthead committed:
fix(test): disabled items properly tested

CataldoMazzilli commented:
@nubsthead @beawar what about renaming selected to selectedItems?  selected appear to be a boolean

CataldoMazzilli commented:
@nubsthead @beawar what about makes items not optional?
In the 99% of cases items prop is used and when you need an empty array I think that is better specify it as prop
I know that it is a BC

nubsthead commented:
> @nubsthead @beawar what about makes items not optional? In the 99% of cases items prop is used and when you need an empty array I think that is better specify it as prop I know that it is a BC

I agree to transform items as a mandatory prop.

> @nubsthead @beawar what about renaming selected to selectedItems? selected appear to be a boolean

I don't know about that. I imagine a boolean to be isSelected. But I would like to hear more opinion on it.

beawar commented:
@CMazzilli @nubsthead Tbh I feel we miss a convention here. If all the boolean were prefixed with is/has then I would be sure selected is not a boolean. At the moment I'm not sure of that, and so selectedItems feels clearer.
If you agree, we can discuss about this internally and then decide if a renaming of the variable is required or not. So for the moment I wouldn't mark this point as a request for changes.

beawar commented:
@CMazzilli
@nubsthead and I discussed about the breaking change and it's actually breaking the current version of the Table. I'm currently working on a refactor of it, which solves this BC nicely, and so we decided to delay the BC after the refactor of the Table component. An additional point in favor of this solution is that we will have a single/simple commit for the BC which will not intact the changes made by Gabriele in this PR, so in case we need to revert the BC, we will revert only tthose two single lines and not everything else.

I'll wait for your approval in order to merge this one.
